Chapter 10: MicroProfile JWT (#25)
microprofile/microprofile-tutorial: chapter10/index.adoc

== Chapter 10: JWT Authentication

In modern microservices architectures, where services are distributed and stateless,
securing communications between clients and services — and between individual services — is critical. JSON Web Token (JWT) provides a lightweight, self-contained and efficient mechanism for user authentication and authorization, enabling scalable and secure identity propagation across distributed systems. 

> MicroProfile JWT is a specification that standardizes JWT-based authentication and authorization for Java microservices. By leveraging the JWT open standard (RFC and seamlessly integrates  with Jakarta EE security to provide role-based access 7519), it enables services to securely extract and validate claims such as identity, roles and permissions.

Reviewer: ```suggestion
MicroProfile JWT is a specification that standardizes JWT-based authentication and authorization for Java microservices. By leveraging the JWT open standard https://datatracker.ietf.org/doc/html/rfc7519[RFC 7519], it enables services to securely extract and validate claims such as identity, roles and permissions.
```

==== What is a JSON Web Token (JWT)?

https://jwt.io/[JSON Web Token (JWT)], defined by https://datatracker.ietf.org/doc/html/rfc7519[RFC 7519] is an open standard for securely transmitting information (claims) between parties as a JSON object. JWTs are digitally signed, ensuring their integrity and authenticity.

=== Structure of a JWT

A JWT consists of three Base64 encoded parts, separated by dots (+.+): 

[source]
----
<Header>.<Payload>.<Signature>
----

- *Header* -  It contains metadata about the token, such as token type (type: “JWT”) and signing algorithm ( alg: “RS256” for RSA-SHA256). 

[source, json]
----
{
  "alg": "RS256",
  "typ": "JWT"
}
----
- *Payload* - It contains claims which are key-value pairs representing data about the user such as roles and expiration.
Example of claims in a JWT payload:

[source, json]
----
{
  "iss": "https://io.microprofile.com/issuer",
  "sub": "user1",
  "exp": 1735689600,
  "iat": 1735686000,
  "aud": "my-audience",
  "groups": ["user", "admin"]
}
----

> - *Signature* - A digital signature that is used to verify the token’s integrity by combining the encoded header, payload, and a secret/private key.

Reviewer: ```suggestion
- *Signature* - A digital signature that is used to verify the token’s integrity by combining the encoded header, payload, and a private key.
```

We do not support symmetric key signatures formally

Example JWT Token:
[source]
----

> eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ1c2VyMTIzIiwiaXNzIjoiaHR0cHM6Ly9hdXRoLmV4YW1wbGUuY29tIiwiZ3JvdXBzIjpbInVzZXIiLCJhZG1pbiJdfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c

Author: Fixed.

----

==== Types of Claims in MicroProfile JWT

Claims in JWTs can be categorized into two types:

===== Standard Claims

> These are predefined claims with specific meanings, as defined by the JWT specification. MicroProfile JWT recognizes and validates these claims during token validation. Some commonly used standard claims include:

Reviewer: ```suggestion
These are predefined claims with specific meanings, as defined by the JWT specification. Some commonly used standard claims include:
```

MP JWT does not currently deal with the `jti` and `nbf` claims.

Perhaps, `MicroProfile JWT recognizes and validates these claims during token validation.` can remain but then please remove these 2 claims from the table below

|===
|*Claim*|*Description*|*Example*

|`iss`|Issuer, the entity that issued the JWT (e.g., an authentication server)|`"iss": "https://io.microprofile.com/issuer"`
|`sub`|Subject, the principal (user or service) that the JWT is about.|`"sub": "user1"`
|`aud`|Audience, the intended recipients of the token (e.g., specific microservices).|`"aud": "order-service"`
|`exp`|Expiration time|`"exp": 1735689600`
|`nbf`|not before|`"nbf": 1735686000`
|`iat`|issued at, when that token was issued.|`"iat": 1735686000`
|`jti`|Unique JWT token identifier|`"jti": "a1b2c3d4"`
|`groups`|Groups, list of roles or users allowed to access the resource |`["user", "admin"]`
|===

===== Custom Claims

These are application-specific claims that provide additional information about the user or entity. They can be used to extend authorization logic with application-specific claims (e.g., `department`, `region`). Custom claims are not part of the JWT specification but are often used to include domain-specific data, such as user preferences, tenant IDs, or other metadata. MicroProfile JWT allows developers to access these claims programmatically.

> === Benefits of JWTs in Microservices?

Reviewer: ```suggestion
=== Benefits of JWTs in Microservices
```

> JWTs are a foundational technology for securing distributed systems due to their compact, self-contained, and stateless nature. These characteristics make them particularly well-suited for microservices architectures, where scalability, interoperability, and decentralized security are critical.

Reviewer: ```suggestion
```

This sounds too complex, example, services are stateless but JWTs can not be described as stateless. Neither they are compact. A simple sentence that you added below is great IMHO

JWTs are widely used in microservices for the following reasons:

==== Statelessness & Scalability

> No Server-Side Sessions: JWTs eliminate the need for centralized session storage. Each token is self-contained, embedding all necessary user claims (e.g., roles, permissions) in its payload.

Reviewer: ```suggestion
JWTs eliminate the need for centralized session storage. Each token is self-contained, embedding all necessary user claims (e.g., roles, permissions) in its payload.
```

Sessions have gone away, but what is indeed the case is that the session cookie with the encrypted JWT token can help those who don't want to keep the server state. The `session` is a confusing term for this tutorial.

Independent Validation: Microservices validate JWTs locally using public keys, avoiding calls to a central authority. This reduces latency and scales horizontally.

Example:

> A payment service validates a JWT’s signature and exp claim without querying an authentication server. +

Reviewer: ```suggestion
A payment service validates a JWT’s signature without querying an authentication server. +
```

==== Interoperability

Open Standards: JWTs adhere to RFC 7519, ensuring compatibility across platforms (Java, .NET, Node.js) and frameworks (Spring Boot, Quarkus).

MicroProfile Integration: MicroProfile JWT standardizes validation and claim extraction, enabling seamless interoperability across Java microservices.

==== Fine-Grained Authorization

> Custom Claims: JWTs support custom claims (e.g., department, tenant_id) for granular access control.

Reviewer: ```suggestion
```

RBAC mentioned below is the key MP JWT offer

Role-Based Access Control (RBAC): Map JWT claims (e.g., groups) to Jakarta EE roles using @RolesAllowed.

==== Decentralized Security

Propagation Across Services: A JWT issued by an authentication service is propagated across microservices (e.g., Order Service, Inventory Service). Each service independently verifies the token and enforces access control.

Reduced Central Dependency: No need for a central authorization server, simplifying architecture and improving 
fault tolerance.

Example:

- Authentication Service: Issues a JWT with `sub: "user1"` and `groups: ["user"]`.

- Order Service: Validates the JWT and processes requests if groups include `users`.

- Inventory Service: Revalidates the same JWT without contacting the auth service.

> ==== Enhanced Performance

Reviewer: I'd just drop this section completely, they are definitely not complact and can easily reach 4K for more complex deployments

Author: Fixed. Dropping this section as advised.

# Expected issuer (e.g., your OIDC provider)  
mp.jwt.verify.issuer=https://auth.example.com  

# Optional: Validate token audience  
mp.jwt.verify.audiences=order-service,payment-service
----

Explanation: 

- The `mp.jwt.verify.publickey.location` property specifies the location of the public key used to verify the JWT’s signature. 

- The `mp.jwt.verify.issuer` property defines the expected issuer of the JWT, ensuring that tokens are only accepted if they are issued by a trusted authority. 

- Optionally, the `mp.jwt.verify.audiences` property can be used to specify the allowed audiences for the JWT, ensuring that the token is intended for the service.

=== Public Key Setup

Place the PEM-encoded public key in _src/main/resources/META-INF/publicKey.pem_. This key is used to verify incoming JWT signatures.

== Use cases for JSON Web Tokens

JWTs are versatile tokens commonly used in modern applications for authentication, where they verify the identity of a user or service; for authorization, where they grant access to resources based on roles or permissions; and for information exchange, where they securely transmit data between parties.

> Below are key scenarios where JWTs shine in microservices environments:

Reviewer: It feels like this section repeats the section above which talks about Why JWTS are popular in microservices. May be these use cases should be moved there

Author: Fixed.

For example, A user authenticates with an Auth Service and receives a JWT. This JWT token grants access to other services such as product catalog or order management system without re-authentication.

=== Authorization (Role-Based Access Control)

JWTs are also used for authorization, enabling fine-grained access control based on user roles or permissions. The JWT payload typically includes a groups or roles claim, which specifies the user’s roles or permissions. For example, a user with the admin role might be allowed to access all resources while a user with the user role might only have access to specific resources. 

MicroProfile JWT integrates seamlessly with Jakarta EE’s `@RolesAllowed` annotation, making it easy to enforce role-based access control (RBAC) in microservices. Role mapping can be configured in _microprofile-config.properties_: 

[source]
----

> mp.jwt.verify.roles=groups 

Reviewer: This is a default claim so no need to configure it

Author: Good catch. Updated with the required details.

----

> === Stateless Session information

Reviewer: Let's drop this section, it can be a bit confusing talking about sessions, MP JWT is not really about sessions

Author: Removed the section about Stateless Session information.

=== Information Exchange

JWTs can securely exchange information between parties. The token payload can include custom claims representing the data being exchanged, such as an order ID or user ID. This makes JWTs useful in scenarios like Single Sign-On (SSO) systems, where information needs to be shared across multiple services.

For example, a JWT might contain an `order_id` claim and a `user_id` claim, which can be used by an order management service to retrieve and display the user’s order details.

=== Federation & Single Sign-On (SSO)

JWTs facilitate identity federation by allowing integration of multiple trusted identity providers (e.g., Active Directory, LDAP) to provide a single sign-on (SSO) experience. In this case, the JWT contains claims representing the user’s identity, which can be used by applications to identify the user and retrieve their profile information.

For example, an enterprise SSO system can issue a JWT that grants access to HR, Payroll, and CRM microservices. MicroProfile JWT validates the token’s iss (issuer) and aud (audience) to enforce trust boundaries.

== Request Flow in MicroProfile JWT

> Understanding how JWTs propagate and are processed in a microservices architecture is critical to implementing secure and scalable authentication. This section explains the lifecycle of a JWT from client to service, including token extraction, validation, and claim usage.

Reviewer: ```suggestion
Understanding how JWTs are propagated and processed in a microservices architecture is critical to implementing secure and scalable authentication. This section explains the lifecycle of a JWT from client to service, including token extraction, validation, and claim usage.
```

> === How JWTs are Propagated in Microservices

Reviewer: ```suggestion
=== How JWTs are propagated in Microservices
```

Author: Using Title Casing. Not making this change.

JWTs are propagated via the `Authorization: Bearer` HTTP header across clients and services. 

==== Client-to-Service

When a client authenticates (e.g., via a login endpoint), it receives a JWT from an authentication service. This token is then included in the header of subsequent requests to microservices. For example, a request header might look like this: 

[source]
----
GET /api/orders HTTP/1.1
Authorization: Bearer eyJhbGciOiJSUzI1NiIs…
----

> ==== Service-to-Service

Reviewer: This example was already covered above, I'm not sure we should go into this kind of scenario here, as it raises the questions of the correct token audiences (did the Inventory service took a token meant for the Payment one) and token exchange

Author: Fixed.

In a multi-service workflow, the initial microservice (e.g., Order Service) forwards the same token to downstream services (e.g., Payment Service or Inventory Service). 

Each service independently validates the JWT, ensuring decentralized and stateless security.

==== Token Extraction

MicroProfile JWT runtime handles token extraction and validation automatically. The token is parsed and validated as follows:

- Header Parsing: The runtime extracts the token from the Bearer schema.

- Decoding: The JWT is split into its header, payload, and signature components.

==== Token Validation
The token validation involves the following steps: 

> - Signature Verification: The public key (configured via mp.jwt.verify.publickey.location) validates the token’s integrity.

Reviewer: ```suggestion
- Signature Verification: The public key validates the token’s integrity.
```

Other options are also possible

. Validate and Sanitize JWT Claims: Validate all claims (e.g., iss, aud) in microprofile-config.properties and Sanitize custom claims before use can prevent injection attacks and misuse of claims.

. Monitor and Log Security Events: Log JWT validation errors, role mismatches, and token expiration events to detect breaches and audit access patterns. Integrate with monitoring tools (e.g., Prometheus, Grafana) to track anomalies. 

Following these steps will help you secure your microservices against the most common attacks.

== Conclusion

MicroProfile JWT offers a standard-based, interoperable approach for securing microservices.It simplifies identity propagation, access control, and stateless security across distributed services. By integrating with Jakarta EE, it enables secure, scalable, and interoperable authentication without session state.

*Further Reading:*

* https://datatracker.ietf.org/doc/html/rfc7519[RFC 7519]
* https://github.com/eclipse/microprofile-jwt-auth[MicroProfile JWT 2.1 Spec]

> Jakarta Security 3.0

Reviewer 2: Missing link.

```suggestion
* https://jakarta.ee/specifications/security/3.0/[Jakarta Security 3.0]
```
